Improve edge evaluator interface to support compound edge evaluator

Levi-Armstrong opened the pull request:
This reduces code duplication between edge evaluators and enables the ability for the compound edge evaluator to work correctly.

marip8 commented on descartes_light/include/descartes_light/interface/edge_evaluator.h:
```diff
@@ -32,13 +32,59 @@ template <typename FloatType>
+      }
+    }
+
+    for (const auto& rung : edges)
+      if (!rung.empty())
+        return true;
```
A vector of vector of edges (i.e. `edges`) is not comprised of rungs, but of vectors of edges between vertices. I would rename your variables: `rung` -> `edges` and `edges` -> `edge_lists`

marip8 commented on descartes_light/include/descartes_light/interface/edge_evaluator.h:
```diff
@@ -32,13 +32,59 @@ template <typename FloatType>
   virtual bool evaluate(const Rung_<FloatType>& from,
                         const Rung_<FloatType>& to,
-                        std::vector<typename LadderGraph<FloatType>::EdgeList>& edges) = 0;
+                        std::vector<typename LadderGraph<FloatType>::EdgeList>& edges)
```
`edges` here is confusing. If I understand this correctly, this function returns a list of valid edges for each vertex (i.e. joint pose) in the `from` rung. This should probably be called something like `edge_lists` instead

marip8 commented on descartes_light/include/descartes_light/interface/edge_evaluator.h:
```diff
@@ -32,13 +32,59 @@ template <typename FloatType>
+  EdgeEvaluator(std::size_t dof) : dof_(dof) {}
   virtual ~EdgeEvaluator() {}
 
   virtual bool evaluate(const Rung_<FloatType>& from,
```
I don't really like the function signature here. Why does this function return false if the two rungs aren't connected by any vertices? It's not like the evaluation itself failed. I would propose returning the vector of edge lists instead; then let the user decide what it means to have a vector of edge lists that is empty.

```c++
std::vector<typename LadderGraph<FloatType>::EdgeList> evaluate(const Rung_<FloatType>& from, const Rung_<FloatType>& to)
```

Levi-Armstrong replied:
> Why does this function return false if the two rungs aren't connected by any vertices? It's not like the evaluation itself failed. I would propose returning the vector of edge lists instead; then let the user decide what it means to have a vector of edge lists that is empty.

This is an interface class that is only to used by the descartes code which keeps the building of the graph cleaner. The user should not directly be calling evaluate or considerEdge.

marip8 replied:
In this case, the "user" I'm referring to is actually us as developers, not necessarily end-users. I think a change to this interface would clear up the ambiguity with `edges` resizing we're discussing [above](https://github.com/swri-robotics/descartes_light/pull/26#discussion_r499662078). It would be really easy to create a helper function that can look through the output list of edges and determine if it is empty or not to replace the boolean check [here](https://github.com/swri-robotics/descartes_light/blob/e3f6d85063fea4ac77305ea4c180e7c151c1c374/descartes_light/include/descartes_light/impl/descartes_light.hpp#L118). This wouldn't introduce much additional code to the graph construction and would improve the clarity of the edge evaluator interface

Levi-Armstrong replied:
Ok, make sense. Lets create an issue on this and make the change in a separate PR. You good with this?

marip8 commented on descartes_light/include/descartes_light/interface/edge_evaluator.h:
```diff
@@ -32,13 +32,59 @@ template <typename FloatType>
+  }
+
+  /**
+   * @brief This evalutes the edge and determines if it should be considered.
```
"if it should be considered" is pretty ambiguous. I would replace with "Determines whether the edge between two vertices is valid and, if so, its cost."

marip8 commented on descartes_light/include/descartes_light/interface/edge_evaluator.h:
```diff
@@ -32,13 +32,59 @@ template <typename FloatType>
+   * @param end The end state of the edge
+   * @return A pair <True/False, Cost>
+   */
+  virtual std::pair<bool, FloatType> considerEdge(const Rung_<FloatType>& from,
```
If this is a new function, you might consider renaming to something more descriptive like `computeEdgeValidity` or something like that. "Consider" seems pretty ambiguous to me

Also, do you really need access to the rungs? It seems like this function really only needs to know about the two vertices and the timing constraint between them.

Levi-Armstrong replied:
> Also, do you really need access to the rungs? It seems like this function really only needs to know about the two vertices and the timing constraint between them.

I originally did not have it but currently the timing constraint is tied to the rung. I would like to get ride of the timing constraint  in the rung because it should just be a separate edge evaluator. You good with removing it from the rung?

marip8 replied:
I think separating the timing constraint from the the rung would be fine. It already comes into the `build` function as a [separate vector](https://github.com/swri-robotics/descartes_light/blob/e3f6d85063fea4ac77305ea4c180e7c151c1c374/descartes_light/include/descartes_light/descartes_light.h#L42) anyway

marip8 commented on descartes_light/include/descartes_light/interface/edge_evaluator.h:
```diff
@@ -32,13 +32,59 @@ template <typename FloatType>
+   * @param start The start state of the edge
+   * @param to The rung associated with the end state
+   * @param end The end state of the edge
+   * @return A pair <True/False, Cost>
```
More description -> True if edge is valid, false otherwise. Cost to move from the first vertex to the second

marip8 commented on descartes_samplers/include/descartes_samplers/evaluators/impl/compound_edge_evaluator.hpp:
```diff
@@ -23,15 +23,28 @@
+  for (auto& evaluator : evaluators)
+  {
+    auto results = evaluator->considerEdge(from, start, to, end);
+    if (results.first)
+      return std::make_pair(false, cost);
```
If the edge is valid, return false?

Levi-Armstrong replied:
Good catch.

marip8 commented on descartes_samplers/include/descartes_samplers/evaluators/impl/distance_edge_evaluator.hpp:
```diff
@@ -85,26 +58,17 @@ bool DistanceEdgeEvaluator<FloatType>::evaluate(const Rung_<FloatType>& from,
-    {
-      const auto* end_vertex = to.data.data() + dof * j;
+    FloatType step = end[i] - start[i];
+    if (std::abs(step) > delta_thresholds[i])
```
~~`step` is a joint distance and `delta_thresholds[i]` a joint velocity, so shouldn't you have to divide `step` by `dt` to compare?~~

Didn't realize that `delta_thresholds` multiplies by the timing constraint

marip8 commented on descartes_samplers/include/descartes_samplers/evaluators/impl/distance_edge_evaluator.hpp:
```diff
@@ -85,26 +58,17 @@ bool DistanceEdgeEvaluator<FloatType>::evaluate(const Rung_<FloatType>& from,
-      const auto* end_vertex = to.data.data() + dof * j;
+    FloatType step = end[i] - start[i];
+    if (std::abs(step) > delta_thresholds[i])
+      return std::make_pair(true, cost);
```
If the velocity required to go from `start` to `end` is greater than the allowed threshold, then shouldn't you return false?

Levi-Armstrong replied:
Good catch.

marip8 commented on descartes_samplers/include/descartes_samplers/evaluators/impl/distance_edge_evaluator.hpp:
```diff
@@ -85,26 +58,17 @@ bool DistanceEdgeEvaluator<FloatType>::evaluate(const Rung_<FloatType>& from,
-
-  for (std::size_t i = 0; i < n_start; ++i)
+  FloatType cost = static_cast<FloatType>(0.0);
+  for (std::size_t i = 0; i < delta_thresholds.size(); ++i)
```
`delta_thresholds` is confusing to me. It's actually the joint distance limit that each joint can travel to meet the timing constraint, so what does delta mean? We should rename this to something like `joint_distance_threshold`

Levi-Armstrong replied:
Yep it is joint distance.

marip8 commented on descartes_samplers/include/descartes_samplers/evaluators/impl/distance_edge_evaluator.hpp:
```diff
@@ -85,26 +58,17 @@ bool DistanceEdgeEvaluator<FloatType>::evaluate(const Rung_<FloatType>& from,
-      // Consider the edge:
-      considerEdge(start_vertex, end_vertex, delta_thresholds, j, edges[i]);
-    }
+    cost += std::pow(step, FloatType(2));
```
What is the difference between this evaluator and the Euclidean distance evaluator? Just the fact that this one takes timing into account?

Levi-Armstrong replied:
Yes this just takes timing into account.

marip8 replied:
Timing really should be a separate edge evaluator like you mentioned above. It's confusing that there are two evaluators with similar names that do the same thing except one takes timing into account

Levi-Armstrong replied:
Done.

marip8 commented on descartes_samplers/include/descartes_samplers/evaluators/impl/gantry_euclidean_distance_edge_evaluator.hpp:
```diff
@@ -25,63 +25,25 @@ DESCARTES_IGNORE_WARNINGS_POP
-    if (!rung.empty())
-      return true;
+  FloatType cost = 0.0;
+  for (std::size_t i = 2; i < this->dof_; ++i)
```
There are a lot of assumptions being made here:
- The gantry has two external axes
- The gantry axes are the first two values in the list
- The joint distance of the gantry axes is not considered

We need to add these assumptions to the documentation at the very least. Ultimately, we should probably parameterize them with the arguments to the constructor.

Why is the motion of the first two axes of the gantry irrelevant here?

Levi-Armstrong replied:
Based on experience it works better to minimize robot motion and let the gantry do most of the work if it can. Though this is old and not used anymore. I good with marking it deprecated or removing it completely.

marip8 replied:
We don't necessarily need to delete it (unless you really think it needs to go); we just need to document these assumptions on the class definition so we can easily remember how it works

Levi-Armstrong replied:
I think it has little value since most will likely use there own because MoveIt and Tesseract have there own way of defining kinematics. Tesseract implements its own which allow any number of gantry joints and I would assume if other motion planning libraries would do the same.

Levi-Armstrong replied:
I update your Issue #5 to include documentation items that need to be addressed.

marip8 reviewed (changes requested):
Nominally I think this is a good change. It's worth thinking about making the two interface function definitions (`evaluate` and `considerEdge`) as simple and intuitive as possible.

We really also need to comment these classes and what they're intended to be used for. There is a lot of ambiguity on the types of costs that these evaluators should return (i.e. L1 or L2 distance) and the assumptions they make (i.e. do they consider timing, axis order, etc.). We shouldn't have to dig into the source code to understand what these classes/functions do

Levi-Armstrong commented on descartes_light/include/descartes_light/interface/edge_evaluator.h:
```diff
@@ -32,13 +32,59 @@ template <typename FloatType>
 
-  typedef typename std::shared_ptr<EdgeEvaluator<FloatType>> Ptr;
+    // Allocate if needed because of Compound Evaluator)
+    if (edges.size() == 0)
+      edges.resize(n_start);
+
+    assert(edges.size() == n_start);
```
You could create a compound edge evaluator that calls evaluate instead of consider edge so you do not want to resize which would nullify the data already in the vector. Should only resize if zero, but I have added the assert to catch if something goes wrong, the assert should not ever be hit but just encase something changes upstream or a user defined edge evaluator inadvertently changes the size to something incorrect.

Levi-Armstrong replied:
I am not sure if this would ever be done but if we want to fix the size then I would remove the virtual.

marip8 replied:
This is why I want to change this function signature: it's unclear what state `edges` should be in when you call this function and then you need to add checks on `edges` in this function to accommodate all the different ways it could come in. I think the function signature change described below would make this better

Levi-Armstrong replied:
@marip8 You good with me removing the virtual and doing as you recommended?

marip8 replied:
I'm on-board with changing this function

Levi-Armstrong replied:
Done.

Levi-Armstrong commented:
@marip8 I made most of the changes but there is a few I commented on to further discuss.

Levi-Armstrong commented:
@johnwason Do you know how to fix the windows build.?

johnwason commented:
It looks like you are using the old ROS Windows repository with the azurewebsites url. They have started using a new one recently. You should upgrade to the latest install instructions:

http://wiki.ros.org/Installation/Windows

I also suggest using the github for building Windows, since they are using powershell instead of git bash for the environment. The ROS Windows has build in powershell support.

Levi-Armstrong commented:
> It looks like you are using the old ROS Windows repository with the azurewebsites url. They have started using a new one recently. You should upgrade to the latest install instructions:
> 
> http://wiki.ros.org/Installation/Windows
> 
> I also suggest using the github for building Windows, since they are using powershell instead of git bash for the environment. The ROS Windows has build in powershell support.

@johnwason Are you currently using this in any of your repository that I may use as a guide?

Levi-Armstrong commented:
@marip8 I believe I have addressed the remaining issues in the latest commit. I removed the timing constraint and simplified the edge evaluator API. I also updated the issue on documentation related to include items that need to be addressed. I may have time in the near future to go through and add documentation to this repository.

marip8 commented:
I think we should handle the timing constraint a little differently. I think we have 3 options:
1. Make timing a standalone evaluator
    - All other edge evaluator implementations should be independent of timing
1. Make timing a parameter of the `considerEdge` interface (with default value = inf)
    - Edge evaluators can choose whether or not they consider timing
    - Con:
      - It's could be unclear to an end-user whether or not timing is being considered if it's not documented well
      - You would need to copy edge evaluators such that one considers timing and the other doesn't
1. Make timing an argument to the constructor of the edge evaluator implementation
    - Cons:
      - You would need to copy edge evaluators such that one considers timing and the other doesn't

I prefer option 1 (standalone timing evaluator) because it can be easily combined with any other evaluator with the `CompoundEdgeEvaluator` and it eliminates ambiguity in the other evaluators about whether or not they consider timing.

johnwason commented:
@Levi-Armstrong Sorry no, I don't have a unit test using ROS Windows.

marip8 commented on descartes_samplers/include/descartes_samplers/evaluators/impl/timing_edge_evaluator.hpp:
```diff
@@ -0,0 +1,57 @@
+template <typename FloatType>
+std::pair<bool, FloatType> TimingEdgeEvaluator<FloatType>::considerEdge(const FloatType* start, const FloatType* end)
+{
+  Eigen::Matrix<FloatType, Eigen::Dynamic, 1> delta =
+      Eigen::Map<const Eigen::Matrix<FloatType, Eigen::Dynamic, 1>>(end, static_cast<Eigen::Index>(this->dof_)) -
+      Eigen::Map<const Eigen::Matrix<FloatType, Eigen::Dynamic, 1>>(start, static_cast<Eigen::Index>(this->dof_));
+  Eigen::Matrix<FloatType, Eigen::Dynamic, 1> joint_times = delta.cwiseQuotient(velocity_limits_).cwiseAbs();
```
FYI you can use `Eigen::Array` instead of `Eigen::Matrix` to do coefficient-wise operations "natively" without calling `cwise...`

Levi-Armstrong replied:
I assumed they did the same thing under the hood. Is there a benefit for one over the other?

marip8 replied:
`Eigen::Matrix` is for linear algebra operations whereas `Eigen::Array` is for coefficient-wise operations. According to the documentation, you can convert one to the other with no run-time cost. I think it's just more convenient to use one over the other depending on what type of operations you're trying to do

johnwason commented:
@ooeygui Can you help Levi with his question about CI with the updated ROS install instructions?

ooeygui commented:
@johnwason Thanks for loop in - I have an email thread with Levi on the CI.

johnwason commented:
@ooeygui I am interested in the solution to this problem as well.
